source/common/upstream/ring_hash_lb.cc

```diff
@@ -60,34 +62,108 @@ HostConstSharedPtr RingHashLoadBalancer::Ring::chooseHost(uint64_t h) const {
 
 using HashFunction = envoy::api::v2::Cluster_RingHashLbConfig_HashFunction;
 RingHashLoadBalancer::Ring::Ring(
-    const absl::optional<envoy::api::v2::Cluster::RingHashLbConfig>& config,
-    const HostVector& hosts) {
+    const HostsPerLocality& hosts_per_locality,
+    const LocalityWeightsConstSharedPtr& locality_weights,
+    const absl::optional<envoy::api::v2::Cluster::RingHashLbConfig>& config) {
   ENVOY_LOG(trace, "ring hash: building ring");
-  if (hosts.empty()) {
-    return;
+
+  // Sanity-check that the locality weights, if provided, line up with the hosts per locality.
+  if (locality_weights != nullptr) {
+    ASSERT(locality_weights->size() == hosts_per_locality.get().size());
   }
 
-  // Currently we specify the minimum size of the ring, and determine the replication factor
-  // based on the number of hosts. It's possible we might want to support more sophisticated
-  // configuration in the future.
   // NOTE: Currently we keep a ring for healthy hosts and unhealthy hosts, and this is done per
   //       thread. This is the simplest implementation, but it's expensive from a memory
   //       standpoint and duplicates the regeneration computation. In the future we might want
   //       to generate the rings centrally and then just RCU them out to each thread. This is
   //       sufficient for getting started.
+
   const uint64_t min_ring_size =
-      config ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, 1024) : 1024;
+      config
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), minimum_ring_size, DefaultMinRingSize)
+          : DefaultMinRingSize;
+  const uint64_t max_ring_size =
+      config
+          ? PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.value(), maximum_ring_size, DefaultMaxRingSize)
+          : DefaultMaxRingSize;
+  const absl::optional<uint64_t> configured_replication_factor =
```

Comment:
Can't this just be a uint and set it to 1 if not specified?

Reply:
Better to leave it as an `optional`, I think - I need to be able to tell the difference between an unspecified replication factor (in which case we use the old behavior for backward compat) and a specified replication factor = 1 (in which case we'll warn the user if what they specified is too small or too large for the ring bounds).